etl/afd_ems_import/incident_import.py

```diff
@@ -0,0 +1,220 @@
+#!/usr/bin/ python
+import csv
+import email
+from io import BytesIO, TextIOWrapper
+import logging
+import os
+import sys
+
+from boto3 import client, resource
+
+from utils.cli import get_cli_args
+from utils.columns import COLUMNS
+from utils.graphql import make_hasura_request
+from utils.queries import MUTATIONS
+
+AWS_ACCESS_KEY_ID = os.getenv("AWS_ACCESS_KEY_ID")
+AWS_SECRET_ACCESS_KEY = os.getenv("AWS_SECRET_ACCESS_KEY")
+BUCKET_NAME = os.getenv("BUCKET_NAME")
+BUCKET_ENV = os.getenv("BUCKET_ENV")
+
+BATCH_SIZE = 1000
+
+s3_client = client("s3")
+s3_resource = resource("s3")
+
+
+logging.basicConfig(stream=sys.stdout, level=logging.INFO)
+
+
+def get_emails_todo(source, subdir="inbox"):
+    """Get a list of S3 object keys of the email messages in the /inbox.
+
+    Args:
+        source (str): The data source: `afd` or `ems`
+        subdir (str, optional): The S3 bucket subdirectory to check. Defaults to "inbox".
+
+    Raises:
+        IOError: If not objects are found in the bucket subdirectory
+
+    Returns:
+        List: List of S3 object keys, sorted oldest to newest by modified date
+    """
+    prefix = f"{BUCKET_ENV}/{source}_incidents/{subdir}"
+    logging.info(f"Checking for files in: {prefix}")
+    response = s3_client.list_objects(
+        Bucket=BUCKET_NAME,
+        Prefix=prefix,
+    )
+    emails = []
+    for item in response.get("Contents", []):
+        key = item.get("Key")
+        # ignore the subdirectory itself
+        if not key.endswith("/"):
+            last_modified = item.get("LastModified")
+            emails.append((key, last_modified))
+    if not len(emails):
+        raise IOError("No emails found in S3 bucket")
+    # sort oldest -> newest
+    emails.sort(key=lambda x: x[1])
+    # return object keys only
+    return [email[0] for email in emails]
+
+
+def download_email(email_obj_key):
+    """Download an email message from S3"""
+    logging.info(f"Downloading: {email_obj_key}")
+    file_object = BytesIO()
+    s3_client.download_fileobj(BUCKET_NAME, email_obj_key, file_object)
+    file_object.seek(0)
+    return file_object.read().decode("utf-8")
+
+
+def get_data_from_email_str(source, email_str):
+    """Parse raw email content and return the attachment as a list of dictionaries
+
+    Args:
+        source (str): The data source: `afd` or `ems`
+        email_str (str): the decoded email string
+
+    Returns:
+        list: list of row dictionaries of the parsed CSV attachment
+    """
+    logging.info("Getting attachment from email...")
+    message = email.message_from_string(email_str)
+    attachment = message.get_payload()[1]
+    attachment_data = BytesIO(attachment.get_payload(decode=True))
+
+    logging.info("Parsing attachment data...")
+    if source == "ems":
```

Comment:
some awkward forking here and elsewhere to do ems- and afd- specific handling :/